journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournalWriter.java

```diff
@@ -80,34 +91,68 @@ public void append(final JournalRecord record) {
     }
   }
 
-  public void reset(final long index) {
-    currentSegment = journal.resetSegments(index);
+  void reset(final long index) {
+    currentSegment = segments.resetSegments(index);
     currentWriter = currentSegment.writer();
+    lastFlushedIndex = index - 1;
   }
 
-  public void deleteAfter(final long index) {
+  void deleteAfter(final long index) {
     // Delete all segments with first indexes greater than the given index.
-    while (index < currentSegment.index() && currentSegment != journal.getFirstSegment()) {
-      journal.removeSegment(currentSegment);
-      currentSegment = journal.getLastSegment();
+    while (index < currentSegment.index() && currentSegment != segments.getFirstSegment()) {
+      segments.removeSegment(currentSegment);
+      currentSegment = segments.getLastSegment();
       currentWriter = currentSegment.writer();
     }
 
     // Truncate the current index.
     currentWriter.truncate(index);
+    lastFlushedIndex = index - 1;
   }
 
-  public void flush() {
-    journalMetrics.observeSegmentFlush(currentWriter::flush);
-  }
+  /**
+   * Fetches all segments with a last index greater than or equal to current {@link
+   * #lastFlushedIndex}. These are then flushed in order. The {@link Segment#lastIndex()} of the
+   * last successful segment to be flushed will be stored in the given {@link JournalMetaStore}.
+   *
+   * @param metaStore where to store the last successfully flushed index
+   */
+  void flush(final JournalMetaStore metaStore) {
+    final var dirtySegments = segments.getSegments(lastFlushedIndex);
+    final int segmentsCount = dirtySegments.size();
+    long flushedIndex = -1;
+
+    if (segmentsCount == 0) {
+      LOGGER.debug(
+          "No segments to flush for last flushed index {} (last index = {})",
+          lastFlushedIndex,
+          getLastIndex());
+      return;
+    }
 
-  public void close() {
-    currentWriter.close();
+    try {
+      for (final var segment : dirtySegments) {
+        final long lastSegmentIndex = segment.lastIndex();
```

Comment:
`segment.lastIndex()` is not accessing a volatile field. Wouldn't that be a problem if flush is executed on another thread? What would happen if it gets an incorrect value?

Reply:
I thought about that. In general it shouldn't matter as you may just end up flushing more data than expected, however if we've truncated or reset it might be a problem, since we would then save that as the last flushed index, which would be wrong, I think. So it might be we do need to ensure it's volatile unfortunately.

I also thought about passing the expected last index during scheduling, but again, this may change due to reset/truncate in between.

Reply:
Looks like flush should be synchronized with `deleteAfter` and `reset`. And after deleteAfter and reset, lastFlushedIndex should be immediately updated. :thinking:

Reply:
I think with the following we can resolve this issue:
- SegmentedJournal:Flush acquire rwLock.readLock before flushing. All delete methods already acquire write lock and it happens only rarely. So it doesn't add any contention.
- SegmentedWriter:deleteAfter and reset update lastFlushedIndex in metastore  and in-memory before delete. No need to update after if it is not flushed. 
- Next step, see if we have to remove the lock when direct flush is configured for perf reasons.

wdyt?

Reply:
@npepinpe Let's sync on this monday. If it is becoming more complex, we can re-think if we really want to do this or not.

Reply:
:+1: 

I didn't have much time to look into this today anyway.

I think the `lastIndex` access here is actually not super critical and there's an easy workaround (i.e. when scheduling the flush, pass the expected last index from the Raft thread as an argument). I think the truncate/reset issues are the more complicated ones.

Reply:
I added acquiring the read lock. It seems to have little impact in general, including in the blocking case. There's little contention for the lock, since it would only content with compact/reset/truncate, which are all done synchronously as well, and it will not contend with the readers.

I would go with this for now, as it will also ease making compaction asynchronous if we want to.

Reply:
:+1: Please ping me when it is ready for review.

Reply:
I started a benchmark with the synchronous flush just to verify my claims, let's see

Reply:
Benchmark with synchronous flush looks comparable to the baseline. Initially I had worse performance because of all the logging I added actually, but that wouldn't be debug in the final PR anyway. Looks fine now

Reply:
Sync flushing with acquiring lock:

![image](https://user-images.githubusercontent.com/43373/220070859-a841ebfb-f77a-4797-93d8-06fbb2faa3a3.png)
![image](https://user-images.githubusercontent.com/43373/220070228-e08c5625-3d58-4107-b7d1-f1187b02cb84.png)
![image](https://user-images.githubusercontent.com/43373/220070322-a46633bc-3d7c-4317-be19-fca157450002.png)
![image](https://user-images.githubusercontent.com/43373/220070366-6c04212b-102a-4ee9-b732-71e5bfe9f966.png)


Without acquiring lock (i.e. baseline):

![image](https://user-images.githubusercontent.com/43373/220070763-183ea024-8ac0-4e6b-a8a8-5de8717357ca.png)
![image](https://user-images.githubusercontent.com/43373/220070645-f61b7546-c039-4364-9932-cdcf35ae1eb3.png)
![image](https://user-images.githubusercontent.com/43373/220070579-a7f2b69c-b0d4-45ad-a4c1-09930512f719.png)
![image](https://user-images.githubusercontent.com/43373/220070518-f21039ca-f68d-4eb0-801b-699866574963.png)

Reply:
One change with the sync flush versus baseline is we don't flush on segment creation anymore. You can see it has a very small improvement there. The p99 is similar, but more of them are clustered in the lower ranges:

Sync flush:

![image](https://user-images.githubusercontent.com/43373/220106416-82d27917-d70d-401d-a767-f31d8c6932a0.png)

Baseline:

![image](https://user-images.githubusercontent.com/43373/220106557-d640cded-eaff-499b-895c-3d09092fc1fa.png)

Reply:
It seems very small though, so I wouldn't immediately classify that as a win, just figured I'd mention it :smile:

Reply:
Great :+1: